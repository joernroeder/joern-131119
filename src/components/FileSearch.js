import React, { useState, useEffect } from 'react'
import {
  useFilteredFilesDispatch,
  Actions,
  useFilteredFilesState,
} from '../store/FilteredFileStore'

import useDebounce from '../hooks/useDebounce'
import { useApiContext } from '../api/ApiContext'

// TODO implement loading indicator and show errors
const FileSearch = () => {
  // debounced value
  const [value, setValue] = useState('')
  const debouncedValue = useDebounce(value, 100)
  // stores value sent to the api
  const [valueSent, setValueSent] = useState('')

  // get query and dispatch from filteredFiles State
  const { query } = useFilteredFilesState()
  const dispatch = useFilteredFilesDispatch()

  const api = useApiContext()
  const { run, cancel } = api.fetchFilteredFiles({
    defer: true,
    onResolve: ({ data }) => {
      dispatch({
        type: Actions.SET_FILTER,
        payload: {
          query: valueSent, // <- use of valueSent
          filteredFiles: data,
        },
      })
    },
  })

  useEffect(() => {
    if (debouncedValue === valueSent) {
      return
    }

    // cancel running requests.
    cancel()

    // resetting query
    if (!debouncedValue && query) {
      dispatch({ type: Actions.DISABLE_FILTER })
      return
    }

    run((init) => {
      return {
        ...init,
        resource: init.resource + encodeURIComponent(debouncedValue),
      }
    })
    setValueSent(debouncedValue)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [debouncedValue, query])

  const onInputChange = (event) => {
    setValue(event.target.value)
    event.preventDefault()
  }

  return (
    <form className="sm:order-1">
      <input
        type="search"
        className="border text-lg w-full px-4 py-1"
        placeholder="Search documentsâ€¦"
        value={value}
        onChange={onInputChange}
      />
    </form>
  )
}

export default FileSearch
