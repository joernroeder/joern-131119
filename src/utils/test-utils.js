import React from 'react'

import '@testing-library/jest-dom/extend-expect'
import 'jest-prop-type-error'
import fetchMock from 'jest-fetch-mock'

import { render } from '@testing-library/react'

import { FilesProvider } from '../store/FileStore'
import { FilteredFilesProvider } from '../store/FilteredFileStore'
import { ApiProvider } from '../api/ApiContext'

global.fetch = fetchMock

const WithFilesProvider = ({ children }) => {
  return (
    <FilesProvider>
      <FilteredFilesProvider>{children}</FilteredFilesProvider>
    </FilesProvider>
  )
}

const WithApiAndFilesProvider = ({ children }) => {
  return (
    <ApiProvider>
      <WithFilesProvider>{children}</WithFilesProvider>
    </ApiProvider>
  )
}

const renderWithFilesProviders = (ui, options) => {
  return render(ui, { wrapper: WithFilesProvider, ...options })
}

const renderWithApiAndFileProviders = (ui, options) => {
  return render(ui, { wrapper: WithApiAndFilesProvider, ...options })
}

// re-export everything
export * from '@testing-library/react'

// override render method
export { renderWithFilesProviders, renderWithApiAndFileProviders }
