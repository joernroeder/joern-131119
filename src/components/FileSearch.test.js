import React from 'react'

import { renderWithApiAndFileProviders, fireEvent, wait } from 'test-utils'

import FileSearch from './FileSearch'

afterEach(() => {
  // cleaning up the mess left behind the previous test
  fetch.resetMocks()
})

test('file search should render correctly', () => {
  const { container } = renderWithApiAndFileProviders(<FileSearch />)

  expect(container).toMatchSnapshot()
})

test('file search should trigger a search on value change', async () => {
  const { queryByPlaceholderText } = renderWithApiAndFileProviders(
    <FileSearch />
  )

  const search = queryByPlaceholderText(/Search documents/)

  expect(search.value).toBe('')
  fireEvent.change(search, { target: { value: 'someInp' } })

  await wait(() => {
    expect(fetch).toHaveBeenCalled()
    const [uri] = fetch.mock.calls[0]

    expect(uri).toContain('/files?q=someInp')

    fetch.mockResponse({
      data: [
        {
          type: 'file',
          id:
            '4c1700116bfc788eca30ab1561a5f7821c686c6302b78e17f107e2c96a59d51c',
          attributes: {
            name: '1552892759_00cd432e47_o',
            size: 1193928,
            mime: 'image/jpeg',
          },
        },
      ],
    })
  })
})

test('file search should send the query parameter encoded', async () => {
  const injectedScript = '<script>alert("here i am!")</script>'

  const { queryByPlaceholderText } = renderWithApiAndFileProviders(
    <FileSearch />
  )

  const search = queryByPlaceholderText(/Search documents/)

  fireEvent.change(search, {
    target: { value: injectedScript },
  })

  await wait(() => {
    console.log(fetch.mock.calls[0])
    const [uri] = fetch.mock.calls[0]

    expect(uri).toContain(
      '?q=%3Cscript%3Ealert(%22here%20i%20am!%22)%3C%2Fscript%3E'
    )
  })
})
