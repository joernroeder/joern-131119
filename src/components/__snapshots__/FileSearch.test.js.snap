// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file search should render correctly 1`] = `
<div>
  <form
    class="sm:order-1"
  >
    <input
      class="border text-lg w-full px-4 py-1"
      placeholder="Search documentsâ€¦"
      type="search"
      value=""
    />
  </form>
</div>
`;
